server/routes/login.ts

```diff
@@ -0,0 +1,167 @@
+    csrfToken: res.locals.csrfToken,
+  }
+
+  logError(req.url, data, 'Login failure')
+
+  res.render('pages/index', data)
+}
+
+export default function loginRouter(): Router {
+  const router = Router()
+  const {
+    notify: { url, clientKey, smsTemplateId, emailTemplateId },
+  } = config
+  const notify = url ? new NotifyClient(url, clientKey) : new NotifyClient(clientKey)
+
+  const notifySmsTemplate = smsTemplateId || ''
+  const notifyEmailTemplate = emailTemplateId || ''
+
+  const postLogin = async (req: Request, res: Response) => {
+    const ipAddress = req.get('x-forwarded-for') || req.socket.remoteAddress || ''
+
+    let userNotSignedUpMessage = false
+
+    try {
+      const userAuthenticationDetails = await userAuthenticationService.getUserAuthenticationDetails(req.user.username)
+
+      if (userAuthenticationDetails === null || userAuthenticationDetails.length === 0) {
+        userNotSignedUpMessage = true
+      }
+
+      const quantumAddresses = config.quantumAddresses.split(',')
+      if (
+        !utilities.hmppsAuthMFAUser(req.user.token) &&
+        config.twoFactorAuthOn === 'true' &&
+        !ipRangeCheck(ipAddress, quantumAddresses)
+      ) {
+        if (userAuthenticationDetails === null || userAuthenticationDetails.length === 0) {
+          processError(
+            new Error(`Error : No Sms or Email address returned for QuantumId : ${req.user.username}`),
+            req,
+            res,
+            userNotSignedUpMessage,
+          )
+        }
+
+        // eslint-disable-next-line no-shadow
+        const userAuthentication = userAuthenticationDetails[0]
+
+        if (
+          (userAuthentication.EmailAddress === null || userAuthentication.EmailAddress === '') &&
+          (userAuthentication.Sms === null || userAuthentication.Sms === '')
+        ) {
+          processError(
+            new Error(`Error : Sms or Email address null or empty for QuantumId : ${req.user.username}`),
+            req,
+            res,
+            userNotSignedUpMessage,
+          )
+        }
+
+        const emailEnabled = userAuthentication.UseEmailAddress
+        const smsEnabled = userAuthentication.UseSms
+
+        if (!emailEnabled && !smsEnabled) {
+          processError(
+            new Error(`Error : Sms or Email address both set to false for QuantumId : ${req.user.username}`),
+            req,
+            res,
+            userNotSignedUpMessage,
+          )
+        }
+
+        const twofactorCode = utilities.get2faCode()
+
+        await userAuthenticationService.updateTwoFactorAuthenticationHash(
+          req.user.username,
+          utilities.createTwoFactorAuthenticationHash(twofactorCode.toString()),
+        )
+
+        if (smsEnabled) {
+          await notify
+            .sendSms(notifySmsTemplate, userAuthentication.Sms || '', {
+              personalisation: { '2fa_code': twofactorCode },
+            })
+            .catch((err: string) => {
```

Comment:
dubious about this, think `err` should be of type `Error` but that's what the existing code is doing